config/azure/build.yaml

```diff
@@ -43,15 +54,32 @@ jobs:
      displayName: 'Install Toolchain (Darwin)'
      condition: eq( variables['Agent.OS'], 'Darwin' )
 
+   - script: g++ --version
+     condition: or( eq( variables['Agent.OS'], 'Darwin' ), eq( variables['Agent.OS'], 'Linux' ) )
+   
+   # Windows needs the vcvarsall.bat command in the same script to retain environment variables for the scons task
    - script: |
-       g++ --version
+       call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
```

Comment:
It's a shame there isn't a way to do this in the same way as `UsePythonVersion` :-( I saw we're potentially not the only ones after this either https://github.com/Microsoft/azure-pipelines-tasks/issues/9737. 

Given we're stuck with this, do you think we could parameterise the test cases so we don't end up accidentally only updating one of them? 

If we did this with a template parameter + default value, would this help with https://github.com/ImageEngine/cortex/pull/994/commits/e56d6b696a87c7e8b87529dc1640b2c9f65bc7a4, as we could then supply a limited list of tests that do work for the windows run - or do too many fail?

If not, it could just be a normal variable.

Reply:
Yeah, this was an annoying one to sort out. Considering it's a Microsoft tool and running Windows I had foolishly assumed they would have it set up out of the box to build Windows programs 😱

A variable sounds good, I can definitely picture myself overlooking the change in multiple spots.

I'm not 100% following you on the 'limited list of tests' part though. Do you mean putting the full command (including which tests to run) in the parameter?

Regarding the tests in general, I have a follow-up PR I need to tidy up that gets about 95% of them running. There are some annoying Windows quirks that cause some tests to give false-negative results (failing due to cleanup after the test) that I want to sort out.

I think I'll opt for now to just put the path of the .bat file in a variable since I'm hoping to have all test passing shortly.

Reply:
Doh! Hadn't realised it was that bad for windows builds out the box! Thanks for all the time wrangling it.

> I'm not 100% following you on the 'limited list of tests' part though. Do you mean putting the full command (including which tests to run) in the parameter?

I'd just noticed that you had disabled the tests all together on Windows. If some worked, then making a template parameter for which tests  to run, would allow `azure-pipelines.yaml` to override the default list for windows, and avoid the duplication in the platform specific tasks in one go.  eg:

```
parameters:
  tests: 'testCore testCorePython testScene testImage testAlembic testAppleseed'
```
and override them in `azure-pipelines.yaml`
```
- template: config/azure/build.yaml
    parameters:
      name: WINDOWS
      displayName: Windows
      vmImage: 'vs2017-win2016'
      options: config/azure/options.windows
      tests: testCore
```
If you can get them all working, thats great.  In Gaffer, we have some `inCI` decorators we wrap around tests we know fail in CI, but in this case, doing it in the template was to avoid the duplication in the script definitions too though.

Reply:
@ericmehl Thinking about it - I like your idea much better - if we made a template param along the lines of `envSetupCmd`, and set it to `call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64` in `azure-pipelines.yaml`, then we could probably get rid of the duplicate tasks all together right?  It would also then allow you to matrix in different VS versions if needed?

Reply:
Sounds like a plan! As long as this currently running build succeeds I'll push over my changes and note the appropriate commits in the thread here.

Reply:
> Sounds like a plan! As long as this currently running build succeeds I'll push over my changes and note the appropriate commits in the thread here.

👍 Thanks for all the time on this!

Reply:
My pleasure! Have a look at 0fdd8a4 - I think it does what we are talking about.